### PR TITLE
Created smurf crate monitor agent

### PR DESCRIPTION
**Pull request by msilvafe**
## Description
I created an agent for monitoring all of the available sensors within a smurf ATCA crate. The agent ssh-s to the shelf manager of the crate who's ip is specified in your site config file using the `subprocess` package. It then takes the following steps:

1. Init function confirms that the docker can communicate with the crate shelf and then proceeds to data acquisition if so or stops and throws and error for debugging if not.
2. The crate is then queeried and the stdout is parsed to make a list of all available sensors. 
3. The list of sensor names are then converted from hardware addresses into human readable feednames which match the influxdb requirements based on the shelf manager standards.
4. Then the acquisition is started, and the shelf is queeried once/30 sec and sensors corresponding to the list assembled during the setup step are assembled into the correct data dictionary format and published.

## Motivation and Context
This will allow us to monitor fan speeds, temperatures, and currents of the smurf crates. It will be useful in the lab but especially important for remote monitoring in the field.

## How Has This Been Tested?
I built a docker image on adaq1 and connected to the crate in k2so. I validated that the feeds show up in grafana and that I can read them back and plot them in simons1.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] Unless I am preparing a release, I have opened this PR onto the `develop` branch.

### CODE REVIEW

**Comment by mhasself** [comment]
Thanks folks -- this is not a full review, just mentioning a couple of things you will definitely need to fix:
- The python code appears to be indented with tabs, not spaces.  Please replace tabs with 4 spaces (and figure out how to change your text editor settings so it doesn't do this again).
- There are two temporary files (*~) committed.  You know what to do.
- In your process loop, you sleep for 30 seconds.  During that 30 seconds, you do not check for the exit condition.  This means that when a user requests the process to stop it could be delayed for up to 30 seconds (and there would still be another read before it exits).  One potential fix would be to move your sleep to the bottom of the loop and turn it into 30 x 1 second sleeps, checking for the exit condition variable on each iteration.

File: docs/agents/smurf_crate_monitor.rst [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+Agent in a docker container.
+
+ocs-config
+``````````
+To configure the SMuRF Crate Monitor Agent we need to add a CrateAgent entry 
+to our site configuration file. Here is an example configuration block using 
+all of the available arguments::
+
+        {'agent-class': 'CrateAgent',
+         'instance-id': 'smurf-crate-monitor',
```
Comment (jlashner) [inline docs/agents/smurf_crate_monitor.rst:36]: I think we should put the crate-id in the instance-id in this example. Something like `crate1-monitor`.

File: docs/agents/smurf_crate_monitor.rst [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+You also need to add the ocs-base anchor and mount the home directory of 
+the ocs-user in your 'docker-compose' file, see the docker section of the
+docs for an example.
+
+The second argument, 'crate-id', is just an identifier for your feed names
+to distinguish between identical sensors on different crates.
+
+Docker
+``````
+The LabJack Agent should be configured to run in a Docker container. An
```
Comment (jlashner) [inline docs/agents/smurf_crate_monitor.rst:69]: smurf_crate agent

File: docs/agents/smurf_crate_monitor.rst [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+2. Switch to ocs user using 'sudo su ocs'
+
+3. 'ssh' into the smurf-crate and add ssh host-verification when prompted
+
+4. Copy ocs-user ssh key using 'ssh-copy-id'
+
+You also need to add the ocs-base anchor and mount the home directory of 
+the ocs-user in your 'docker-compose' file, see the docker section of the
+docs for an example.
```
Comment (jlashner) [inline docs/agents/smurf_crate_monitor.rst:62]: can you add a link here or copy what the yaml anchor looks like? I'm not actually sure the ocs-base tag is described anywhere in the docs.

File: docs/agents/smurf_crate_monitor.rst [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+	ocs-smurf-crate-monitor:
+	<<: *ocs-base
+	image: ocs-smurf-crate-monitor:latest
+	hostname: adaq1-docker
+	network_mode: "host"
+	volumes:
+	  - ${OCS_CONFIG_DIR}:/config
+	  - /home/ocs:/home/ocs
+	command:
+	  - "--instance-id=smurf-crate-monitor"
```
Comment (jlashner) [inline docs/agents/smurf_crate_monitor.rst:81]: indent

**Review by jlashner — CHANGES_REQUESTED** [review]
Thanks max! Looks really good. I added some comments inline, mostly cosmetic/pep8 and docs related.

File: agents/smurf_crate_monitor/smurf_crate_monitor.py [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+    def start_acq(self, session, params=None):
+        """Starts acquiring data, hardcoded for one data point every 30
+        seconds because we intend for this to be very low rate data."""
+        print('Started acquisition')
+        ipmbs, sensids, chan_names = self.init_data_stream(session = session,
+                                                        shm_addr = self.shm_addr)
+        print('Got sensor names')
+        self.take_data = True
+        while self.take_data:
+            print('Take Data = True')
```
[the remark this answers is not in the document]
Comment (msilvafe) [reply agents/smurf_crate_monitor/smurf_crate_monitor.py]: Removed

File: agents/smurf_crate_monitor/smurf_crate_monitor.py [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+            for _ in range(30):
+				if not self.take_data:
+					break
+				time.sleep(1)
```
[the remark this answers is not in the document]
Comment (msilvafe) [reply agents/smurf_crate_monitor/smurf_crate_monitor.py]: Fixed...got these settings fixed in emacs on my laptop, adaq, and smurf server too.

File: agents/smurf_crate_monitor/smurf_crate_monitor.py [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+            datadict = self.get_data_dict(session = session, 
+                                            shm_addr = self.shm_addr,
+                                            ipmbs = ipmbs, 
+                                            sensids = sensids, 
+                                            chan_names = chan_names,
+                                            crate_id = self.crate_id)
+            print('Collected Data Packet')
```
[the remark this answers is not in the document]
Comment (msilvafe) [reply agents/smurf_crate_monitor/smurf_crate_monitor.py]: Removed in [commit](https://github.com/simonsobs/socs/pull/139/commits/2a3418542f31b6f7353349ceb750db6b40efae6b)

File: agents/smurf_crate_monitor/smurf_crate_monitor.py [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+            data = {
+                    'timestamp': time.time(),
+                    'block_name': f'smurf_{self.crate_id}',
+                    'data': datadict
+                    }
+            self.agent.publish_to_feed(f'{self.crate_id}_sensors',data)
+            print('Published Data Packet')
```
[the remark this answers is not in the document]
Comment (msilvafe) [reply agents/smurf_crate_monitor/smurf_crate_monitor.py]: Removed in [commit](https://github.com/simonsobs/socs/pull/139/commits/2a3418542f31b6f7353349ceb750db6b40efae6b)

File: agents/smurf_crate_monitor/smurf_crate_monitor.py [hunk 9]
```diff
@@ -0,0 +1,274 @@
...
+        self.agent = agent
+        self.log = agent.log
+        self.shm_addr = shm_addr
+        self.crate_id = crate_id
+        # Register feed
+        agg_params = {
+            'frame_length': 10*60
+        }
+        print('registering')
+        self.agent.register_feed(f'{crate_id}_sensors',
```
[the remark this answers is not in the document]
Comment (msilvafe) [reply agents/smurf_crate_monitor/smurf_crate_monitor.py]: Changed it to follow your naming suggestion and tested that the feed names updated in grafana.

File: agents/smurf_crate_monitor/smurf_crate_monitor.py [hunk 10]
```diff
@@ -0,0 +1,274 @@
...
+                                 record=True,
+                                 agg_params=agg_params,
+                                 buffer_time=0.)
+                                 
+    def get_sensors(self, session, shm_addr):
```
[the remark this answers is not in the document]
Comment (msilvafe) [reply agents/smurf_crate_monitor/smurf_crate_monitor.py]: Got it, removed the session parameter from all of those functions and function calls.

File: docs/agents/smurf_crate_monitor.rst (hunk 3 above)
Comment (msilvafe) [reply docs/agents/smurf_crate_monitor.rst:62]: I just added a section with an example of the ocs-base anchor here.

**Review by BrianJKoopman — CHANGES_REQUESTED** [review]
This'll be great, thanks for writing the Agent Max. Some inline comments mostly on docs and logging, and some more general/structural ones here:
- The docs page should be added to the main `index.rst` file, in the alphabetical list under "Agent Reference".
- Structure wise, there isn't really a need to have `get_sensors`, `get_channel_names`, or `get_data_dict` be methods within the Agent class, none of them use `self`. They should be moved to just be functions. This would change some of the recs I've made inline for using `self.log`, you should follow the recommendations here: https://ocs.readthedocs.io/en/develop/developer/agents.html#logging
- The reason your early [builds failed](https://github.com/simonsobs/socs/runs/1900787362?check_suite_focus=true) is that you were referencing a function that didn't exist to build the arguments list in Sphinx. You need to pull out the parser creation like in other Agents, i.e. https://github.com/simonsobs/socs/blob/135df3bc08c7c9745d9291bc1d72e6c29feff7eb/agents/smurf_recorder/smurf_recorder.py#L164

**Review by jlashner — APPROVED** [review]
Alright, looks good to me once Brian's changes are addressed. Thanks max!

**Comment by msilvafe** [comment]
@BrianJKoopman and @jlashner ok I think I addressed all comments and checks are passing now.

File: docs/agents/smurf_crate_monitor.rst (hunk 4 above)
Comment (BrianJKoopman) [reply docs/agents/smurf_crate_monitor.rst]: This last thing still isn't resolved, should indent lines 73-81. @msilvafe

Comment (BrianJKoopman) [reply docs/agents/smurf_crate_monitor.rst]: Ah, must have been looking at an outdated view in my browser, looks fixed. Marked as resolved.

**Review by BrianJKoopman — APPROVED** [review]
Looks good, thanks for the fast turn around, Max!

**Pull request merged**